src/base/Flow.ts

```diff
@@ -0,0 +1,314 @@
+
+  /**
+   * Set the basic options, initialize queues and redis client
+   * @param options parameters of AccessFlow
+   */
+  constructor(options: FlowOptions) {
+    const { name, logger, redisClientOptions, queueOptions, lookupAttributes } =
+      options;
+
+    this.name = name;
+    this.logger = logger;
+    this.lookupAttributes = lookupAttributes;
+    this.redisClientOptions = redisClientOptions;
+    this.redis = createClient(redisClientOptions);
+
+    this.queues = queueOptions.map((qo) => new Queue(qo));
+  }
+
+  /**
+   * Connect to redis client
+   */
+  private connect = async () => {
+    await this.redis.connect();
+  };
+
+  /**
+   * Disconnect from redis client
+   */
+  private disconnect = async () => {
+    await this.redis.disconnect();
+  };
+
+  /**
+   * Create a job and put it in the first queue
+   * @param options parameters of the flow
+   * @returns the job's id
+   */
+  public createJob = async (options: CreateJobOptions): Promise<string> => {
+    // generate id for the job
+    const jobId = uuidV4();
+    const jobKey = `${JOB_KEY_PREFIX}:${this.name}:${jobId}`;
+
+    const transaction = this.redis
+      .multi()
+      // create the job with the parameters
+      .hSet(jobKey, objectToStringEntries(options));
+
+    // add lookup keys
+    this.lookupAttributes.forEach((la) => {
+      if (typeof options[la] === "string" || typeof options[la] === "number") {
+        // if attribute is primitive add one key
+        transaction.rPush(
+          `${JOB_KEY_PREFIX}:${this.name}:${la}:${options[la]}`,
+          jobId
+        );
+      } else if (options[la] instanceof Array) {
+        // if it's an array, add one for each element
+        options[la].forEach((iterator: any) => {
+          transaction.rPush(
+            `${JOB_KEY_PREFIX}:${this.name}:${la}:${iterator}`,
+            jobId
+          );
+        });
+      }
+    });
+
+    // put to the first queue
+    transaction.rPush(this.queues[0].waitingQueueKey, jobId);
+
+    // execute transaction
+    await transaction.exec();
+
+    return jobId;
+  };
+
+  /**
+   * Get jobs by ids
+   * @param jobIds job ids
+   * @param resolveChildren whether include child jobs (not just their keys)
+   * @returns jobs
+   */
+  private getJobs = async (jobIds: string[], resolveChildren: boolean) => {
+    const transaction = this.redis.multi();
+    jobIds.forEach((jobId) => {
+      const jobKey = `${JOB_KEY_PREFIX}:${this.name}:${jobId}`;
+      transaction.hGetAll(jobKey);
+    });
+    const jobStrings = await transaction.exec();
+    let jobs = jobStrings.map((j) => parseObject(j as any));
+
+    if (resolveChildren) {
+      // yes, we need this many awaits here
+      jobs = await Promise.all(
+        jobs.map(async (j) =>
+          Object.fromEntries(
+            await Promise.all(
+              Object.entries(j).map(async ([key, value]) => {
+                if (key.match(/^children:.*:jobs$/) && value instanceof Array) {
```

Comment:
else throw error?

Reply:
We don't need to. If the key is not a "children" key, we just return the original key-value pair since it don't need to be resolved.